Add regex_replace tool dependency action.

davebx opened the pull request:
This obviates the pull requests I made to tools-iuc and tools-devteam, but of course the shell commands would need to be replaced with this new action type.

davebx committed:
Add regex_replace tool dependency action.

bgruening commented on lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py:
```diff
@@ -872,6 +873,52 @@ def prepare_step( self, tool_dependency, action_elem, action_dict, install_envir
+
+    def prepare_step( self, tool_dependency, action_elem, action_dict, install_environment, is_binary_download ):
+        '''
+        Populate action_dict with the provided filename, regex, and replacement text.
```
Can we add an example XML snippet here to have this documented?

yhoogstrate replied:
Don't forget to copy the example to the wiki by the time it gets merged:
https://wiki.galaxyproject.org/ToolDependenciesTagSets

bgruening commented on lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py:
```diff
@@ -872,6 +873,52 @@ def prepare_step( self, tool_dependency, action_elem, action_dict, install_envir
+            filename = os.path.abspath( os.path.join( current_dir, action_dict[ 'filename' ] ) )
+        regex = re.compile( action_dict[ 'regex' ], flags=re.MULTILINE )
+        replacement = action_dict[ 'replacement' ]
+        haystack = file( filename, 'r' ).read()
```
Maybe a `for match in pattern.finditer(text):` would be more secure for large files?

davebx replied:
Actually, since I'm using re.MULTILINE here, it should probably iterate through the file line by line.

hexylena replied:
any reason for not just calling out to sed?
- would convert more cleanly from the existing usages 
  - (though no one uses it in really strange ways, so I guess that's not a real 'pro')
- likely faster
  - (though file sizes are tiny, so I suppose it doesn't matter in this usage)

bgruening replied:
@davebx but you are loading the entire file into memory, right?
@erasche using python gives us a nice abstraction between different implementations, the actual reason why this feature was requested.

davebx replied:
@erasche Because sanitizing user-provided input to a shell command is intensely annoying, and it's nearly impossible to cover everything.
@bgruening Yes. I just tested with `for line in fh.readlines()`, which also loaded the entire file into memory. If I iterate until eof with `readline()`, a 1.8GB file consumes a maximum memory of 6.8MB. I'll be updating this pull request with the latter implementation.

hexylena replied:
alright, sounds good. @davebx I didn't see this as "use-provided" shell input nearly as much...I mean, they have `<action type="shell_command">`, not like they need another avenue for doing shell commands ;)

davebx replied:
@erasche Point taken. That said, my general preference is to err on the side of being more careful than necessary.

hexylena replied:
works for me :)

tir. 14. jul. 2015 kl. 12.59 skrev Dave B. notifications@github.com:

> In lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py
> https://github.com/galaxyproject/galaxy/pull/457#discussion_r34598497:
> 
> > -        filtered_actions and dir.
> > -        """
> > -        log_file = os.path.join( install_environment.install_dir, basic_util.INSTALLATION_LOG )
> > -        if os.path.exists( log_file ):
> > -            logfile = open( log_file, 'ab' )
> > -        else:
> > -            logfile = open( log_file, 'wb' )
> > -        if os.path.isabs( action_dict[ 'filename' ] ):
> > -            filename = action_dict[ 'filename' ]
> > -            if not ( filename.startswith( current_dir ) or filename.startswith( install_environment.install_dir ) ):
> > -                return tool_dependency, None, None
> > -        else:
> > -            filename = os.path.abspath( os.path.join( current_dir, action_dict[ 'filename' ] ) )
> > -        regex = re.compile( action_dict[ 'regex' ], flags=re.MULTILINE )
> > -        replacement = action_dict[ 'replacement' ]
> > -        haystack = file( filename, 'r' ).read()
> 
> @erasche https://github.com/erasche Point taken. That said, my general
> preference is to err on the side of being more careful than necessary.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/galaxyproject/galaxy/pull/457/files#r34598497.

davebx committed:
Add documentation to the execute_step method.

davebx committed:
Improved memory usage for regex_replace action.

yhoogstrate commented on lib/tool_shed/galaxy_install/tool_dependencies/recipe/step_handler.py:
```diff
@@ -872,6 +874,79 @@ def prepare_step( self, tool_dependency, action_elem, action_dict, install_envir
+        line = 'placeholder'
+        total_replacements = 0
+        with open( filename, 'r' ) as haystack:
+            while len( line ) > 0:
```
I think `for line in haystack:` would be a more 'Pythonic' way to do this. If I see it correct it will give the same result and allows you to get rid of `line = 'placeholder'`.

davebx committed:
Minor cleanup to regex_replace.

- Iterate through the input file in a more pythonic way.
- Flush the output file, just in case.

davebx commented:
@yhoogstrate Good point, I've updated the method.

bgruening commented:
LGTM! @davebx do we have any test framework for this?

davebx commented:
@bgruening Not yet. I'll add that to the pull request, ideally today.

jmchilton commented:
This wouldn't be the only tag without tests, I would say we should just merge this before it becomes stale if the tests are taking awhile to produce/not a priority.

:+1:

bgruening commented:
Ok ;)

bgruening commented:
@davebx are you adding this to the wiki?
